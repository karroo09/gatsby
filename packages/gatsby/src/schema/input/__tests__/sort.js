const { TypeComposer } = require(`graphql-compose`)
const {
  GraphQLInputObjectType,
  GraphQLEnumType,
  GraphQLBoolean,
  GraphQLList,
} = require(`graphql`)

const getSortInput = require(`../sort`)
const { addNodeInterface } = require(`../../interfaces`)

const tc = TypeComposer.create({
  name: `Foo`,
  fields: {
    bool: `Boolean`,
    array: [`Boolean`],
    nested: [`type Nested { bool: [Boolean!]! }`],
    enum: `enum CustomEnum { FOO BAR }`,
  },
})
addNodeInterface(tc)
const itc = tc.getITC()
const sort = getSortInput(itc)

describe(`Sort input`, () => {
  it(`builds sort input type`, () => {
    expect(sort.getType()).toBeInstanceOf(GraphQLInputObjectType)
    expect(sort.getFieldNames()).toEqual([`fields`, `order`])
  })

  it(`does not mutate input`, () => {
    expect(itc.getFieldType(`bool`)).toBe(GraphQLBoolean)
    expect(itc.getFieldType(`array`)).toBeInstanceOf(GraphQLList)
    expect(itc.getFieldType(`array`).ofType).toBe(GraphQLBoolean)
  })

  it(`adds sort order enum`, () => {
    expect(sort.getFieldType(`order`)).toBeInstanceOf(GraphQLEnumType)
    expect(sort.getField(`order`).defaultValue).toBe(`ASC`)
    expect(sort.getFieldType(`order`).getValues()).toEqual([
      expect.objectContaining({ name: `ASC`, value: `ASC` }),
      expect.objectContaining({ name: `DESC`, value: `DESC` }),
    ])
  })

  it(`adds sort fields`, () => {
    expect(sort.getFieldType(`fields`).ofType).toBeInstanceOf(GraphQLEnumType)
    expect(sort.getFieldType(`fields`).ofType.getValues()).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ name: `BOOL`, value: `bool` }),
        expect.objectContaining({ name: `ARRAY`, value: `array` }),
        expect.objectContaining({ name: `ENUM`, value: `enum` }),
        expect.objectContaining({
          name: `NESTED___BOOL`,
          value: `nested.bool`,
        }),
      ])
    )
  })

  it(`adds Node interface fields`, () => {
    const sortKeys = sort.getFieldType(`fields`).ofType.getValues()
    expect(sortKeys).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ name: `ID`, value: `id` }),
        expect.objectContaining({ name: `PARENT___ID`, value: `parent.id` }),
        expect.objectContaining({
          name: `PARENT___PARENT___ID`,
          value: `parent.parent.id`,
        }),
        expect.objectContaining({
          name: `PARENT___PARENT___PARENT___ID`,
          value: `parent.parent.parent.id`,
        }),
      ])
    )
  })

  it(`respects max sort depth`, () => {
    const sortKeys = sort.getFieldType(`fields`).ofType.getValues()
    expect(sortKeys).not.toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          name: `PARENT___PARENT___PARENT___PARENT___ID`,
          value: `parent.parent.parent.parent.id`,
        }),
      ])
    )
  })

  it(`built sort input fields match snapshot`, () => {
    expect(
      sort
        .getFieldType(`fields`)
        .ofType.getValues()
        .map(({ name, value }) => ({ name, value }))
    ).toMatchSnapshot()
  })
})
