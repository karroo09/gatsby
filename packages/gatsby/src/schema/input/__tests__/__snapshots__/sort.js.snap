// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort input built sort input fields match snapshot 1`] = `
Array [
  Object {
    "name": "BOOL",
    "value": "bool",
  },
  Object {
    "name": "ARRAY",
    "value": "array",
  },
  Object {
    "name": "NESTED___BOOL",
    "value": "nested.bool",
  },
  Object {
    "name": "ENUM",
    "value": "enum",
  },
  Object {
    "name": "ID",
    "value": "id",
  },
  Object {
    "name": "PARENT___ID",
    "value": "parent.id",
  },
  Object {
    "name": "PARENT___PARENT___ID",
    "value": "parent.parent.id",
  },
  Object {
    "name": "PARENT___PARENT___PARENT___ID",
    "value": "parent.parent.parent.id",
  },
  Object {
    "name": "PARENT___PARENT___CHILDREN___ID",
    "value": "parent.parent.children.id",
  },
  Object {
    "name": "PARENT___PARENT___INTERNAL___TYPE",
    "value": "parent.parent.internal.type",
  },
  Object {
    "name": "PARENT___CHILDREN___ID",
    "value": "parent.children.id",
  },
  Object {
    "name": "PARENT___CHILDREN___PARENT___ID",
    "value": "parent.children.parent.id",
  },
  Object {
    "name": "PARENT___CHILDREN___CHILDREN___ID",
    "value": "parent.children.children.id",
  },
  Object {
    "name": "PARENT___CHILDREN___INTERNAL___TYPE",
    "value": "parent.children.internal.type",
  },
  Object {
    "name": "PARENT___INTERNAL___TYPE",
    "value": "parent.internal.type",
  },
  Object {
    "name": "CHILDREN___ID",
    "value": "children.id",
  },
  Object {
    "name": "CHILDREN___PARENT___ID",
    "value": "children.parent.id",
  },
  Object {
    "name": "CHILDREN___PARENT___PARENT___ID",
    "value": "children.parent.parent.id",
  },
  Object {
    "name": "CHILDREN___PARENT___CHILDREN___ID",
    "value": "children.parent.children.id",
  },
  Object {
    "name": "CHILDREN___PARENT___INTERNAL___TYPE",
    "value": "children.parent.internal.type",
  },
  Object {
    "name": "CHILDREN___CHILDREN___ID",
    "value": "children.children.id",
  },
  Object {
    "name": "CHILDREN___CHILDREN___PARENT___ID",
    "value": "children.children.parent.id",
  },
  Object {
    "name": "CHILDREN___CHILDREN___CHILDREN___ID",
    "value": "children.children.children.id",
  },
  Object {
    "name": "CHILDREN___CHILDREN___INTERNAL___TYPE",
    "value": "children.children.internal.type",
  },
  Object {
    "name": "CHILDREN___INTERNAL___TYPE",
    "value": "children.internal.type",
  },
  Object {
    "name": "INTERNAL___TYPE",
    "value": "internal.type",
  },
]
`;
